Assets/Scripts/Augment/GunController.cs

```diff
@@ -33,18 +33,42 @@ public class GunController : MonoBehaviour
     public delegate void GunEvent(GunStats gunStats);
 
     public GunEvent onReload;
+    public GunEvent onFireStart;
     public GunEvent onFire;
+    public GunEvent onFireEnd;
     public GunEvent onInitializeGun;
     public GunEvent onInitializeBullet;
 
+    private AugmentAnimator barrelAnimator;
+
+    private bool isFiring = false;
+
+    private void Start()
+    {
+        var barrel = GetComponentInChildren<GunBarrel>();
+        barrelAnimator = barrel.GetComponentInChildren<AugmentAnimator>();
+        if (HasRecoil)
+            barrelAnimator.OnShotFiredAnimation += PlayRecoil;
+        barrelAnimator.OnShotFiredAnimation += ShotFired;
+        barrelAnimator.OnAnimationEnd += FireEnd;
+    }
+
+    private void OnDestroy()
+    {
+        if (HasRecoil)
+            barrelAnimator.OnShotFiredAnimation -= PlayRecoil;
+        barrelAnimator.OnShotFiredAnimation -= ShotFired;
+        barrelAnimator.OnAnimationEnd -= FireEnd;
+    }
+
     private void FixedUpdate()
     {
         if (fireRateController == null)
         {
             // No fireratecontroller exists
             return;
         }
-        if (fireRateController.shouldFire(triggerPressed, triggerHeld))
+        if (!isFiring && fireRateController.shouldFire(triggerPressed, triggerHeld))
```

Comment:
I managed to enter a state in which I couldn't fire the game anymore using DDR body + shotgun (seemingly because I reloaded at a bad timing, it's a bit hard to replicate)

Reply:
Oh dear god why :)))))

Reply:
Unable to reproduce, could you test again?

Reply:
I just managed to reproduce it:
Sometimes we get an VFXformatter out of range exception in BulletController (Which is an old bug).
Before this change, when this exceptions occurs, it seems the gun shot trails aren't rendered properly in that shot.
With this change, since the compiler exits the function before it's finished when the exception occurs, the isFiring is never changed back, permanently disabling the shotgun (or probably any bullet based barrel)

Reply:
FIXED THE INDEX THING